### PR TITLE
api: Add java.time.Duration overloads to CallOptions, AbstractStub

### PR DESCRIPTION
**Pull request by SreeramdasLavanya**
Overloading API's containing long, TimeUnit with java.time.Duration  

Fixes  https://github.com/grpc/grpc-java/issues/10245

### CODE REVIEW

File: api/src/context/java/io/grpc/Deadline.java [hunk 1]
```diff
@@ -95,6 +101,11 @@ public static Deadline after(long duration, TimeUnit units, Ticker ticker) {
     return new Deadline(ticker, units.toNanos(duration), true);
   }
 
+  public static Deadline after(Duration duration, Ticker ticker) {
+    return after(TimeUnit.NANOSECONDS.convert(duration.getSeconds(), TimeUnit.SECONDS),
```
Comment (kannanjgithub) [inline api/src/context/java/io/grpc/Deadline.java:105]: I have a concern about the loss of precision, I have raised a question on Eric's comment to use TimeUnit.NANOSECONDS.convert(duration) instead of Duration.toNano().

Comment (kannanjgithub) [reply api/src/context/java/io/grpc/Deadline.java]: As Eric replied, you should be using https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/TimeUnit.html#convert(java.time.Duration) and not have to convert using the seconds value.

File: api/src/main/java/io/grpc/SynchronizationContext.java [hunk 2]
```diff
@@ -149,16 +151,16 @@ public final ScheduledHandle schedule(
       final Runnable task, long delay, TimeUnit unit, ScheduledExecutorService timerService) {
     final ManagedRunnable runnable = new ManagedRunnable(task);
     ScheduledFuture<?> future = timerService.schedule(new Runnable() {
-        @Override
```
Comment (kannanjgithub) [inline api/src/main/java/io/grpc/SynchronizationContext.java:152]: Don't format lines unrelated to your changes, it will make git blame difficult to identify changes specific to a fix.

File: api/src/test/java/io/grpc/SynchronizationContextTest.java [hunk 3]
```diff
@@ -72,7 +73,7 @@ public void uncaughtException(Thread t, Throwable e) {
 
   @Mock
   private Runnable task3;
-
```
Comment (kannanjgithub) [inline api/src/test/java/io/grpc/SynchronizationContextTest.java:75]: Remove formatting changes in all unrelated places.

File: stub/src/main/java/io/grpc/stub/AbstractStub.java [hunk 4]
```diff
@@ -149,6 +151,10 @@ public final S withDeadlineAfter(long duration, TimeUnit unit) {
     return build(channel, callOptions.withDeadlineAfter(duration, unit));
   }
 
+  public final S withDeadlineAfter(Duration duration) {
```
Comment (kannanjgithub) [inline stub/src/main/java/io/grpc/stub/AbstractStub.java:154]: Missing unit test.

Comment (SreeramdasLavanya) [reply stub/src/main/java/io/grpc/stub/AbstractStub.java:154]: Included required unit test. Requesting you to check.

File: api/src/test/java/io/grpc/InternalTimeUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+package io.grpc;
+
+import static io.grpc.InternalTimeUtils.convert;
+import static org.junit.Assert.assertEquals;
+
+import java.time.Duration;
+import org.junit.Test;
+
+public class InternalTimeUtilsTest {
```
Comment (kannanjgithub) [inline api/src/test/java/io/grpc/InternalTimeUtilsTest.java:25]: This is not how you write unit tests, it does not test InternalTimeUtils in anwyway. You need to call convert(Duration) on InternalTimeUtils and assert for the expected responses.

File: api/src/main/java/io/grpc/SynchronizationContext.java (hunk 2 above)
Comment (kannanjgithub) [reply api/src/main/java/io/grpc/SynchronizationContext.java:152]: The formatting changes have not been reverted, here and in other places.

File: api/src/test/java/io/grpc/InternalTimeUtilsTest.java (hunk 5 above)
Comment (kannanjgithub) [reply api/src/test/java/io/grpc/InternalTimeUtilsTest.java:25]: How is this test testing InternalTimeUtils at all, when you haven't used that class?

Comment (kannanjgithub) [reply api/src/test/java/io/grpc/InternalTimeUtilsTest.java:25]: my bad, by oversight I didn't notice the static import for convert is from InternalTimeUtils class. I was only referring to InternalTimeUtilsTest and it is actually fine

File: api/src/test/java/io/grpc/SynchronizationContextTest.java (hunk 3 above)
Comment (kannanjgithub) [reply api/src/test/java/io/grpc/SynchronizationContextTest.java:75]: Can you take care of this comment? @SreeramdasLavanya

Comment (kannanjgithub) [reply api/src/test/java/io/grpc/SynchronizationContextTest.java:75]: Whitespace formatting changes are still there in this fie.

File: api/src/main/java/io/grpc/InternalTimeUtils.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import java.time.Duration;
+
+@Internal
+public final class InternalTimeUtils {
+  public static long convert(Duration duration) {
```
Comment (kannanjgithub) [inline api/src/main/java/io/grpc/InternalTimeUtils.java:23]: Code coverage is not detecting that this method is used from tests for some reason. I don't know if it is unable to detect static import of a function. Can you try removing the static import of InternalTimeUtils.convert in the test class, and import the class only, and use InternalTimeUtils.convert( in the test code and see if the Codecov warning goes away.

File: api/src/main/java/io/grpc/InternalTimeUtils.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import java.time.Duration;
+
+@Internal
```
Comment (kannanjgithub) [inline api/src/main/java/io/grpc/InternalTimeUtils.java:21]: @Internal is used to mark certain classes as should not be used by user code. This class being just a util helper, there is no need to mark it as Internal. Also drop the Internal prefix from the class and file name and make it just TimeUtils.

**Review by kannanjgithub — APPROVED** [review]

File: api/src/main/java/io/grpc/TimeUtils.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import java.time.Duration;
+
+public final class TimeUtils {
```
Comment (ejona86) [inline api/src/main/java/io/grpc/TimeUtils.java:21]: Don't create a new public API. This should be package-private.

File: api/src/main/java/io/grpc/TimeUtils.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+  public static long convert(Duration duration) {
```
Comment (ejona86) [inline api/src/main/java/io/grpc/TimeUtils.java:22]: It is not at all clear you get nanos back. `convertToNanos()`? `toNanos()`?

File: api/src/main/java/io/grpc/TimeUtils.java (hunk 8 above)
Comment (kannanjgithub) [reply api/src/main/java/io/grpc/TimeUtils.java:21]: We are using it from a different package io.grpc.stub.AbstractStub too. We'd have to make a copy of this class then. Is that what we want? @ejona86

File: api/src/main/java/io/grpc/TimeUtils.java [hunk 10]
```diff
@@ -19,7 +19,7 @@
 import java.time.Duration;
 
 public final class TimeUtils {
-  public static long convert(Duration duration) {
+  public static long convertToNanos(Duration duration) {
     try {
```
Comment (kannanjgithub) [inline api/src/main/java/io/grpc/TimeUtils.java:23]: @ejona86 Why is Codecov complaining even though the class is in fact covered by unit tests?

File: api/src/main/java/io/grpc/TimeUtils.java (hunk 8 above)
Comment (ejona86) [reply api/src/main/java/io/grpc/TimeUtils.java:21]: We have a way to expose internal classes/methods from api:
1. The class/method itself is package-private
2. Create a new class for the class name that needs exposing, prefixing the class name with `Internal` and marking it `@Internal`
3. Add accessors in that `InternalFoo` class.

So here you'd create a public `InternalTimeUtils` class that has `convert()` on it. We do the dance because 1) we hide `Internal*` classes from the javadoc, and 2) when we compile with Blaze we build the `Internal*` classes as a separate target which is visibility-restricted.

We normally do that for APIs needed to coordinate though, not utilities (core has most utilities). In this case, just duplicating the utility seems reasonable. We'll have at most three copies of it: api, stub, core. Most other things depend on core and it it can be public in core. Not the greatest, but it is small.

File: api/src/main/java/io/grpc/InternalTimeUtils.java (hunk 7 above)
Comment (ejona86) [reply api/src/main/java/io/grpc/InternalTimeUtils.java:21]: Well, this is `public` in a public package, so the `@Internal` is needed.

File: api/src/main/java/io/grpc/InternalTimeUtils.java (hunk 6 above)
Comment (ejona86) [reply api/src/main/java/io/grpc/InternalTimeUtils.java:23]: No, it isn't the static import. The code coverage is powered by jacoco which instruments the bytecode to see when a line is executed.

What's probably happening here is it because called in the stub/ tests, but those tests don't instrument api/ classes. I wouldn't worry about it.

**Review by ejona86 — COMMENTED** [review]
We need `@ExperimentalApi` on all the new public API methods (those receiving Duration).

File: api/src/main/java/io/grpc/InternalTimeUtils.java [hunk 11]
```diff
@@ -20,6 +20,7 @@
 
 @Internal
 public final class InternalTimeUtils {
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/10245")
```
Comment (ejona86) [inline api/src/main/java/io/grpc/InternalTimeUtils.java:23]: `@Internal` trumps `@ExperimentalApi`, so no need for it here.

File: api/src/main/java/io/grpc/LoadBalancer.java [hunk 12]
```diff
@@ -156,11 +156,11 @@ public String toString() {
   private int recursionCount;
 
   /**
-   * Handles newly resolved server groups and metadata attributes from name resolution system.
```
Comment (ejona86) [inline api/src/main/java/io/grpc/LoadBalancer.java:159]: Revert these changes.

**Review by ejona86 — APPROVED** [review]

File: api/src/main/java/io/grpc/TimeUtils.java (hunk 10 above)
Comment (kannanjgithub) [reply api/src/main/java/io/grpc/TimeUtils.java:23]: @SreeramdasLavanya can you add the a private constructor?

File: api/src/main/java/io/grpc/SynchronizationContext.java [hunk 13]
```diff
@@ -162,6 +164,12 @@ public String toString() {
     return new ScheduledHandle(runnable, future);
   }
 
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/10245")
```
Comment (ejona86) [inline api/src/main/java/io/grpc/SynchronizationContext.java:167]: These links should point to a new issue created to track stabilizing this API. (See issues with the "experimental API" label for examples.)
